syntax = "proto3";

/*
 This container message is send from Hi Level to Low level and may contain any allowed message in that direction.
*/
message HiToLo {
 oneof payload {
    // Common/configuration messages
    Enable enable = 6;
    Disable disable = 7;
    KeepAliveHi keep_alive = 13;
    FirmwareUpdate firmware_update = 16;
    SetPWM set_pwm = 20;
    AllowPowerOn allow_power_on = 21;
    SetOutputVoltageCurrent set_output_voltage_current = 50;
    Reset reset = 23;
    Replug replug = 24;
  }
}

message AllowPowerOn {
  bool p = 1;
}

message SetOutputVoltageCurrent {
  float voltage = 1;
  float current = 2;
}

message SetPWM {
  int32 mode = 1; // 0: OFF, 1: ON, 2: F
  float duty_cycle = 2;
}

message FirmwareUpdate {
  bool invoke_rom_bootloader = 1;
}

message KeepAliveHi {
  uint32 time_stamp = 1;
  uint32 hw_type = 2;
  uint32 hw_revision = 3;
  uint32 protocol_version_major = 4;
  uint32 protocol_version_minor = 5;
  string sw_version_string = 6;

}

message Replug {
  uint32 time = 1;
}

message Enable { }
message Disable { }
message Reset { }